### PR TITLE
Second home fix Vitalik Ischenko

### PR DESCRIPTION
**Pull request by futurestark**

### CODE REVIEW

**Comment by futurestark** [comment]
also delete garrbage

File: secondHomeAdvanced.cpp [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+    for (auto ch : expression) {
+        switch (ch) {
+        case '(': 
+        case '<': 
+        case '[': 
+        case '{': 
+            stack.push(ch);
+            break;
+        case ')': 
+            if (stack.empty() || stack.top() != '(') return false;
```
Comment (smay1613) [inline secondHomeAdvanced.cpp:28]: use braces even for one-line expression:

if (expr) {
  //action
}

File: secondHomeAdvanced.cpp [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+Немного поучитесь перекидывать с одного контейнера данные в другой:
+2. std::vector<std::pair<std::string>,
+                       size_t> wordCounter (const std::string& text);
+*/
+
+bool isExpressionValid (const std::string& expression)   
+{
+    std::stack<char> stack;
+
+    for (auto ch : expression) {
```
Comment (smay1613) [inline secondHomeAdvanced.cpp:19]: for basic types, it doesn't matter if you use copy (plain auto), but for objects it is critical, so please use const auto& whenever possible

File: secondHomeAdvanced.cpp [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+        case '[': 
+        case '{': 
+            stack.push(ch);
+            break;
+        case ')': 
+            if (stack.empty() || stack.top() != '(') return false;
+            stack.pop();
+            break;
+        case '>': 
+            if (stack.empty() || stack.top() != '<') return false;
```
Comment (smay1613) [inline secondHomeAdvanced.cpp:32]: what if you'll use map for pairs of open and close brackets?

File: secondHomeAdvanced.cpp [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+{
+    if (root == nullptr)
+        return;
+
+    std::list<Node*> queue;
+    queue.push_back(root);
+
+    std::queue<int> fifoQueue;
+
+    Node* curr = nullptr;
```
Comment (smay1613) [inline secondHomeAdvanced.cpp:59]: _naming, use full names_

File: test/ExampleTests.cpp [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+   
+    Node* binaryTree = nullptr;
+    
+    binaryTree = new Node(15);
+    binaryTree->left = new Node(10);
+    binaryTree->right = new Node(20);
+    binaryTree->left->left = new Node(8);
+    binaryTree->left->right = new Node(12);
+    binaryTree->right->left = new Node(16);
+    binaryTree->right->right = new Node(25);
```
Comment (smay1613) [inline test/ExampleTests.cpp:61]: where is the delete? you'll have memory leak here.

File: test/ExampleTests.cpp [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+
+    linkLevelNodes(binaryTree);
+
+    //EXPECT_EQ(false, expresionIsValid)  << std::endl;
```
Comment (smay1613) [inline test/ExampleTests.cpp:65]: and what the check?

**Pull request closed without merging**